cpp/serve/grammar/grammar_state_matcher.cc

```diff
@@ -0,0 +1,403 @@
+ * B: (rule R, choice 0, element 1)
+ * C: (rule S, choice 1, element 0)
+ * D: (rule R, choice 0, element 2)
+ * E: (rule R, choice 1, element 1)
+ * < means the stack top pointers in the previous step.
+ * The stacks in the previous step is: (A, B, C), (A, D), (A, E)
+ *
+ * ### Current step
+ * Current accepted string: abc
+ * Current stack tree:
+ * A-----------------      G<<
+ * |     \     \     \
+ * B---   D<    E<    H
+ * |   \              |
+ * C<   F<<           I<<
+ *
+ * F: (rule S, choice 1, element 1)
+ * G: (rule main, choice 0, element 2) (means the matching process has finished, and will be deleted
+ * when next char comes)
+ * H: (rule R, choice 1, element 2)
+ * I: (rule T, choice 0, element 0)
+ * << means the stack top pointers in the current step.
+ * The stacks in the current step is: (A, B, F), (A, H, I), (G,)
+ */
+
+using namespace tvm::runtime;
+
+TVM_REGISTER_OBJECT_TYPE(GrammarStateMatcherNode);
+
+/* \brief The concrete implementation of GrammarStateMatcherNode. */
+class GrammarStateMatcherNodeImpl : public GrammarStateMatcherNode, public GrammarStateMatcherBase {
+ private:
+  using RuleExpr = BNFGrammarNode::RuleExpr;
+  using RuleExprType = BNFGrammarNode::RuleExprType;
+
+ public:
+  GrammarStateMatcherNodeImpl(std::shared_ptr<GrammarStateInitContext> init_ctx,
+                              int max_rollback_tokens = 0)
+      : GrammarStateMatcherBase(init_ctx->grammar),
+        init_ctx_(init_ctx),
+        max_history_size_(max_rollback_tokens + 1) {}
+
+  bool AcceptToken(int32_t token_id) final {
+    CHECK(init_ctx_->codepoint_tokens_lookup.count(token_id) > 0);
+    const auto& token = init_ctx_->codepoint_tokens_lookup[token_id].token;
+    for (auto codepoint : token) {
+      if (!AcceptCodepoint(codepoint, false)) {
+        return false;
+      }
+    }
+    token_size_history_.push_back(token.size());
+    if (token_size_history_.size() > max_history_size_) {
+      DiscardEarliestSteps(token_size_history_.front());
+      token_size_history_.pop_front();
+    }
+  }
+
+  void FindNextTokenBitmask(NDArray* next_token_bitmask) {
+    const auto& tokens_sorted_by_codepoint = init_ctx_->tokens_sorted_by_codepoint;
+    const auto& catagorized_tokens_for_grammar = init_ctx_->catagorized_tokens_for_grammar;
+    const auto& latest_stack_tops = stack_tops_history_.GetLatest();
+
+    // For every stack, we will either find all tokens it accepts, or all tokens it rejects.
+    // The final rejected tokens should be not accepted by any stack, and also rejected by every
+    // stack.
+
+    // Per stack temporary data.
+    // Note here indices store the indices in tokens_sorted_by_codepoint, instead of the token ids.
+    std::vector<int32_t> accepted_indices;
+    // {-1} means the universal set, i.e. all tokens
+    std::vector<int32_t> rejected_indices{-1};
+    std::vector<int32_t> accepted_indices_delta;
+    std::vector<int32_t> rejected_indices_delta;
+    DynamicBitSet unc_tokens;
```

Comment:
directly construct Bitset by taking view from DLTensor* next_token_bitmask, do not allocate vector here